rotkehlchen/balances/historical.py

```diff
@@ -94,68 +122,341 @@ def get_assets_amounts(
+                    event.timestamp
+                    if isinstance(event, Trade)
+                    else ts_ms_to_sec(event.timestamp)
+                )
+                amounts[event_ts] += current_balances[asset]
+
+        return amounts, negative_balance_event
+
+    def get_historical_netvalue(
+            self,
+            from_ts: Timestamp,
+            to_ts: Timestamp,
+    ) -> tuple[dict[Timestamp, FVal], list[tuple[str, Timestamp]], tuple[str | int | None, str | None] | None]:  # noqa: E501
+        """Calculates historical net worth per day within the given time range.
 
-            if direction == EventDirection.IN:
-                total_amount += event.balance.amount
-            else:  # EventDirection.OUT
-                total_amount -= event.balance.amount
+        Uses asset balances per day combined with historical prices to calculate
+        the total net worth in the user's profit currency for each day. Stops calculation
+        if negative balance would be created for any asset.
 
-            amounts[ts_ms_to_sec(event.timestamp)] = total_amount
-            if total_amount < ZERO:
-                last_event_id = event.event_identifier
+        May raise:
+            - NotFoundError if no events exist in the specified time period.
+            - DeserializationError if there is a problem deserializing an event or trade from DB.
+
+        Returns:
+            - A mapping of timestamps to net worth in main currency for each day
+            - A list of (asset id, timestamp) tuples where price data was missing
+            - The event_identifier that caused a negative balance if any, else None
+        """
+        events, main_currency = self._get_events_and_currency(from_ts=from_ts, to_ts=to_ts)
+        if len(events) == 0:
+            raise NotFoundError(f'No historical data found within {from_ts=} and {to_ts=}.')
+
+        negative_balance_event = None
+        current_balances: dict[Asset, FVal] = defaultdict(FVal)
+        daily_balances: dict[Timestamp, dict[Asset, FVal]] = {}
+        current_day = timestamp_to_daystart_timestamp(
+            events[0].timestamp
+            if isinstance(events[0], Trade)
+            else ts_ms_to_sec(events[0].timestamp),
+        )
+        daily_balances[current_day] = current_balances.copy()
+        for event in events:
+            if (day_ts := timestamp_to_daystart_timestamp(
+                event.timestamp
+                if isinstance(event, Trade)
+                else ts_ms_to_sec(event.timestamp),
+            )) > current_day:
+                daily_balances[current_day] = current_balances.copy()
+                current_day = day_ts
+
+            if (event_ids := self._update_balances(event=event, current_balances=current_balances)) is not None:  # noqa: E501
+                negative_balance_event = event_ids
                 break
 
-        return amounts, last_event_id
+        if negative_balance_event is None:
```

Comment:
add this as an else to the above for loop. With a comment explaining that no negative balance was encountered